core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,333 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Menu;
+
+use Contao\CoreBundle\Event\MenuEvent;
+use Contao\CoreBundle\Monolog\ContaoContext;
+use Contao\CoreBundle\Routing\Page\PageRegistry;
+use Contao\CoreBundle\Security\Authentication\Token\TokenChecker;
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\Database;
+use Contao\Date;
+use Contao\Environment;
+use Contao\PageModel;
+use Contao\StringUtil;
+use Doctrine\DBAL\Connection;
+use Knp\Menu\FactoryInterface;
+use Knp\Menu\ItemInterface;
+use Knp\Menu\MenuItem;
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\Exception\ExceptionInterface;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+class FrontendMenuBuilder
+{
+    private FactoryInterface $factory;
+    private RequestStack $requestStack;
+    private EventDispatcherInterface $dispatcher;
+    private Connection $connection;
+    private PageRegistry $pageRegistry;
+    private TokenChecker $tokenChecker;
+    private Security $security;
+    private LoggerInterface $logger;
+
+    public function __construct(FactoryInterface $factory, RequestStack $requestStack, EventDispatcherInterface $dispatcher, Connection $connection, PageRegistry $pageRegistry, TokenChecker $tokenChecker, Security $security, LoggerInterface $logger)
+    {
+        $this->factory = $factory;
+        $this->requestStack = $requestStack;
+        $this->dispatcher = $dispatcher;
+        $this->connection = $connection;
+        $this->pageRegistry = $pageRegistry;
+        $this->tokenChecker = $tokenChecker;
+        $this->security = $security;
+        $this->logger = $logger;
+    }
+
+    public function getMenu(int $pid, int $level = 1, string $host = null, array $options = []): ItemInterface
+    {
+        $root = $this->factory->createItem('root');
+
+        if (null === ($pages = $this->getPages($pid, $options))) {
+            return $root;
+        }
+
+        $request = $this->requestStack->getCurrentRequest();
+        $requestPage = $request->attributes->get('pageModel');
```

Comment:
I wonder if `$requestPage` should be part of the method signature…

Reply:
I don't think that's needed. It can be achieved by push/pop on the request stack if really needed. Otherwise we could also argue why the logged-in status is not an argument :)